src/sshfs_mount/sftp_server.cpp

```diff
@@ -622,124 +604,122 @@ int mp::SftpServer::handle_open(sftp_client_message msg)
-    const auto len = std::min(msg->len, max_packet_size);
+    const auto& [path, file] = *handle;
 
-    std::vector<char> data;
-    data.reserve(len);
-
-    if (!MP_FILEOPS.seek(*file, msg->offset))
+    if (MP_FILEOPS.lseek(file, msg->offset, SEEK_SET) == -1)
     {
-        mpl::log(mpl::Level::trace, category,
-                 fmt::format("{}: cannot seek to position {} in \'{}\'", __FUNCTION__, msg->offset, file->fileName()));
+        mpl::log(mpl::Level::trace,
+                 category,
+                 fmt::format("{}: cannot seek to position {} in '{}'", __FUNCTION__, msg->offset, path.string()));
         return reply_failure(msg);
     }
 
-    auto r = MP_FILEOPS.read(*file, data.data(), len);
-    if (r < 0)
-    {
-        mpl::log(mpl::Level::trace, category,
-                 fmt::format("{}: read failed for {}: {}", __FUNCTION__, file->fileName(), file->errorString()));
-        return sftp_reply_status(msg, SSH_FX_FAILURE, file->errorString().toStdString().c_str());
-    }
+    constexpr auto max_packet_size = 65536u;
+    std::array<char, max_packet_size> buffer{};
+
+    if (const auto r = MP_FILEOPS.read(file, buffer.data(), std::min(msg->len, max_packet_size)); r > 0)
+        return sftp_reply_data(msg, buffer.data(), r);
     else if (r == 0)
         return sftp_reply_status(msg, SSH_FX_EOF, "End of file");
 
-    return sftp_reply_data(msg, data.data(), r);
+    mpl::log(mpl::Level::trace,
+             category,
+             fmt::format("{}: read failed for '{}': {}", __FUNCTION__, path.string(), std::strerror(errno)));
+    return sftp_reply_status(msg, SSH_FX_FAILURE, std::strerror(errno));
 }
 
 int mp::SftpServer::handle_readdir(sftp_client_message msg)
 {
-    auto dir_entries = handle_from(msg, open_dir_handles);
-    if (dir_entries == nullptr)
+    const auto handle = get_handle<DirIterator>(msg);
+    if (handle == nullptr)
     {
         mpl::log(mpl::Level::trace, category, fmt::format("{}: bad handle requested", __FUNCTION__));
         return reply_bad_handle(msg, "readdir");
     }
 
-    if (dir_entries->empty())
+    auto& dir_iterator = *handle;
+
+    if (!dir_iterator.hasNext())
         return sftp_reply_status(msg, SSH_FX_EOF, nullptr);
 
     const auto max_num_entries_per_packet = 50ll;
-    const auto num_entries = std::min(dir_entries->size(), max_num_entries_per_packet);
-
-    for (int i = 0; i < num_entries; i++)
+    for (int i = 0; i < max_num_entries_per_packet && dir_iterator.hasNext(); i++)
     {
-        const QFileInfo entry = dir_entries->takeFirst();
-        const auto filename = entry.fileName().toStdString();
+        const auto& entry = dir_iterator.next();
+        QFileInfo file_info{entry.path().string().c_str()};
         sftp_attributes_struct attr{};
-        if (entry.isSymLink())
+        if (entry.is_symlink())
         {
-            mp::platform::symlink_attr_from(entry.absoluteFilePath().toStdString().c_str(), &attr);
+            std::error_code _;
```

Comment:
This `_` is not used. I think.

Reply:
True. Must've forgotten it there from a previous attempt.